Check if GEOPM_PLUGIN_PATH is secure

bgeltz opened the pull request:
- Fixes #2229.

cmcantalupo commented on service/geopmdpy/service.py:
```diff
@@ -57,6 +59,20 @@ def __init__(self):
+            user_paths = plugin_path.split(':')
+            for path in user_paths:
+                if system_files.is_secure_path(path):
+                    for file_path in glob.glob(os.path.join(path, '*')):
```
at least currently we also have .json files in the plugin directory.  We don't want to error when these .json files don't end in .so.1.0.0.

cmcantalupo replied:
```suggestion
                    for file_path in glob.glob(os.path.join(path, 'lib*.so*')):
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -14,7 +14,6 @@
 import shutil
 import psutil
 import uuid
-import gi
```
Was this important for any reason?  Seems unnecessary.

bgeltz commented on tutorial/iogroup/README.md:
```diff
@@ -197,6 +197,15 @@ filename must begin with "libgeopmiogroup_" and end in
 An alternative is to install the plugin by copying the .so file into
 the GEOPM install directory, <GEOPM_INSTALL_DIR>/lib/geopm.
 
+> [!WARNING]
```
GitHub markdown now supports admonition:
https://stackoverflow.com/a/72327818

Rendering:
https://github.com/bgeltz/geopm/tree/issue-2229/tutorial/iogroup

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
+            secure_make_dirs(path)
+            scanned_paths = set()
+            # Custom IOGroup/Agent plugins: geopm_plugin.cpp -> plugin_load()
+            for file_path in glob.glob(os.path.join(path, 'lib*.so*')):
```
we have a more strict filter for loading plugins libgeopmpi*.so if i remember.

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
+                    with open(file_path, 'r') as ff:
+                        scanned_paths.add(file_path)
+                        if is_secure_file(file_path, ff):
+                            so_suffix = '.so.' + version.__abi_version__.replace(':', '.')
```
I guess these file checks are better than not having them, but there is a TOCTOU issue with this solution.

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -270,6 +271,42 @@ def is_secure_file(path, fid):
+    if plugin_path is not None:
+        user_paths = plugin_path.split(':')
+        for path in user_paths:
+            secure_make_dirs(path)
```
potentially moves files that are tracked by an RPM or deb file to /usr/lib64/geopm-INVALID, not sure if this is desired, as they will not be removed with an uninstall.

bgeltz committed:
Add docu node about plugin building nuances

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Print geopmread errors during service startup

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo commented on service/src/Helper.cpp:
```diff
@@ -269,6 +271,46 @@ namespace geopm
+        return stat_struct.st_mode;
+    }
+
+    bool is_fd_secure(const int fd, const std::string &plugin) {
```
maybe a more specific name is required for Helper.hpp, e.g. `is_plugin_secure`.

bgeltz commented on service/src/MSRIOGroup.cpp:
```diff
@@ -1260,7 +1261,20 @@ namespace geopm
+                        do_insert = false;
+                    }
+                    else {
+                        // TODO: Insert fd into tracking vector to close on destruction
```
This TODO is actually more difficult than I want it to be.  This function is static and is used in geopmadmin to generate the msr-safe allowlist.  Not sure if it's even worth it as the fd's will be close when the process ends.

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -910,6 +914,8 @@ def _get_names(exec_name):
                                    stdout=subprocess.PIPE,
                                    stderr=subprocess.PIPE)
     stdout_bytes, stderr_bytes = signals_pid.communicate()
```
check return code for subprocess.

bgeltz replied:
Open an issue @bgeltz

cmcantalupo commented on service/src/Helper.cpp:
```diff
@@ -269,6 +271,46 @@ namespace geopm
+                          << plugin << std::endl;
+            }
+            int mode = geopm::fd_to_mode(fd);
+            if (mode != (S_IFREG | S_IRWXU)) { // 0o700
```
Default plugins should have permissions 755 like other files in /usr/lib64.  So the validation here must be that the write bits are not set. For example 755 is okay, 700 is okay, but 777 is not okay

cmcantalupo commented on service/src/MSRIOGroup.cpp:
```diff
@@ -1256,13 +1257,25 @@ namespace geopm
-                std::string filename = dir + "/" + file;
-                if (string_begins_with(file, "msr_") && string_ends_with(file, ".json")) {
-                    data_files.insert(filename);
+            try {
```
```
class SecurePath {
    SecurePath(const std::string &orig_path)
    : SecurePath(orig_path, 0o022, getuid() == 0)
    {}
    SecurePath(const std::string &orig_path, int umask, bool is_same_user);
    ~SecurePath()
    std::string secure_path();
    int fid();
}
```

cmcantalupo replied:
SecurePath sp("foobar");
// replace all uses of "foobar" with sp.secure_path()

bgeltz committed:
Add SecurePath object

- Responsible for opening files and providing
  paths to the open file under procfs.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Use SecurePath for plugin loading

- Add try/catch around directory globbing code for plugins.
- Allows for bad entries in the GEOPM_PLUGIN_PATH
  without preventing future plugins from being
  globbed properly.
- Allows MSRIOGroup to load OK if above.
- Always warn if a plugin failed to load.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz commented on service/src/SecurePath.cpp:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    SecurePath::SecurePath(const std::string &orig_path, mode_t umask, bool enforce)
+        : m_fd(open(orig_path.c_str(), O_RDONLY))
```
Could add O_NOFOLLOW here if we want to disallow symlinks.

open() follows symlinks by default, and fstat cannot tell the difference between a link and a regular file when doing things this way.

bgeltz commented on service/src/geopm_plugin.cpp:
```diff
@@ -77,32 +77,41 @@ namespace geopm
                 if (dl_handle != nullptr) {
                     DLRegistry::dl_registry().add(dl_handle);
                 }
-#ifdef GEOPM_DEBUG
```
Note I'm removing the ifdef.  I think it's important to print into the journal when plugins fail to load unexpectedly.

bgeltz commented on service/docs/source/GEOPM_CXX_MAN_PluginFactory.3.rst:
```diff
@@ -206,6 +206,15 @@ detailed example of plugin load order can be found in
+    LD_LIBRARY_PATH, the plugin will fail to load.
+    For this reason, it is recommended that any plugins are compiled with
+    system installed gcc/g++ compilers.
+
```
Add blurb about not allowing symlinks.

bgeltz commented on tutorial/iogroup/README.md:
```diff
@@ -197,6 +197,15 @@ filename must begin with "libgeopmiogroup_" and end in
+> LD_LIBRARY_PATH, the plugin will fail to load.
+> For this reason, it is recommended that any plugins are compiled with
+> the system installed gcc/g++ compilers.
+
```
Add blurb about not allowing symlinks.

bgeltz commented on service/test/SecurePathTest.cpp:
```diff
@@ -0,0 +1,82 @@
+    chmod(m_file_name.c_str(), test_perms);
+
+    // Assert write permissions for the group/world are not set
+    SecurePath* sp = new SecurePath(m_file_name, (S_IWGRP | S_IWOTH), true);
+    EXPECT_EQ("/proc/self/fd/3", sp->secure_path());
+    delete sp;
```
Try with brackets instead of new/delete.

cmcantalupo replied:
```suggestion
    {
        SecurePath sp(m_file_name, (S_IWGRP | S_IWOTH), true);
        EXPECT_EQ("/proc/self/fd/3", sp.secure_path());
    }
```

cmcantalupo commented on service/src/SecurePath.cpp:
```diff
@@ -0,0 +1,85 @@
+            struct stat stat_struct;
+            int err = fstat(m_fd, &stat_struct);
+            if (err) {
+                (void)close(m_fd);
```
```suggestion
                (void)!close(m_fd);
```
We need to put the ! in there with all these ignored return codes now

cmcantalupo reviewed (commented):
Noticed the missing `!` from the ignored return values on this pass.  Also just wondering why the scoping trick doesn't enable us to avoid the new/delete pattern?
